### DISCUSSION TITLE
How can I avoid ReflectionBinding when binding to a ItemsControls DataContext property from within the Item Template?

### DISCUSSION THREAD

**Discussion by pixsperdavid** [AvaloniaUI/Avalonia#10549] [Q&A] [answered]
A very common pattern from all my previous WPF projects was an items control (of any kind), in which the item template contained buttons which bind to command properties on the main view model. I'm having difficulty finding a good pattern for expressing this without resorting to a ReflectionBinding as shown below:

https://github.com/pixsperdavid/avalonia_binding_example/blob/41e648eb81cf1761617656be7358ce778e77b19c/src/BindingExample/MainWindow.axaml#L38-L39

Is there a way I can express this whilst still being able to use a Compiled Binding and all the benefits that provides? The only other solution I can think of currently is to wrap my `DataItem` class in a view model and include a reference to the parent view model as a property in that, but it adds too much extra complexity to be worth it IMO.

**Comment by timunie** [comment, marked as answer]
Yes, you only need to specify the DataType. Either in your DataTemplate or use x:DataType on any control level. Moreover you can cast in bindings. 

https://docs.avaloniaui.net/docs/data-binding/compiledbindings

**Reply by pixsperdavid** [reply]
So I tried expressing the binding like this
```
Command="{Binding DataContext.DeleteItem, DataType={x:Type local:MainWindowViewModel}, 
								    RelativeSource={RelativeSource AncestorType=ItemsControl}}"
```
but this isn't correct (and doesn't compile), because the source of my binding is the ItemsControl itself, which is not of type MainWindowViewModel.

I also can't set the x:DataType because the button command parameter has to bind to the Item itself in order to pass it to the command.

**Reply by timunie** [reply]
You have to cast the Binding in this case. Also notice that you can do this

x:DataType is not valid in a Binding. Casting works similar to how it's done in C#

**Reply by pixsperdavid** [reply]
Aha, got it, many thanks. The casting was what I was missing as I wasn't aware of this feature. For reference:

```
Command="{Binding ((local:MainWindowViewModel)DataContext).DeleteItem, RelativeSource={RelativeSource AncestorType=ItemsControl}}"
```
